gdnative-core/src/export/method.rs

```diff
@@ -282,18 +282,132 @@ impl<'a> Varargs<'a> {
     pub unsafe fn from_sys(num_args: libc::c_int, args: *mut *mut sys::godot_variant) -> Self {
         let args = std::slice::from_raw_parts(args, num_args as usize);
         let args = std::mem::transmute::<&[*mut sys::godot_variant], &[&Variant]>(args);
-        Self {
-            idx: 0,
-            iter: args.iter(),
-        }
+        Self { idx: 0, args }
     }
 }
 
 impl<'a> Iterator for Varargs<'a> {
     type Item = &'a Variant;
     #[inline]
     fn next(&mut self) -> Option<Self::Item> {
-        self.iter.next().copied()
+        if self.idx < self.args.len() {
+            let idx = self.idx;
+            self.idx += 1;
+            Some(self.args[idx])
+        } else {
+            None
+        }
+    }
+}
+
+// Return a second token.
+macro_rules! replace_expr {
+    ($_t:tt $sub:expr) => {
+        $sub
+    };
+}
+
+// Count parameters.
+macro_rules! count_tts {
+    ($($tts:tt)*) => {
+        0usize $(+ replace_expr!($tts 1usize))*
+    };
+}
+
+// Convert from Varargs to tuples, implement traits.
+macro_rules! varargs_into_tuple {
+    ($($params:ident),*) => {
+        impl<'a, $($params: FromVariant),*> std::convert::TryFrom<Varargs<'a>> for ($($params,)*) {
+            type Error = VarargsError;
+
+            #[inline]
+            fn try_from(value: Varargs<'a>) -> Result<Self, Self::Error> {
+                const EXPECTED: usize = count_tts!($($params)*);
+                let args = &value.args;
+                if args.len() == EXPECTED {
+                    #[allow(unused_variables, unused_mut)]
+                    let mut i: usize = 0;
+                    #[allow(clippy::eval_order_dependence)]
+                    Ok((
+                        $({
+                            i += 1;
+                            $params::from_variant(args[i - 1])?
+                        },)*
+                    ))
+                } else {
+                    Err(ArgumentLengthError{
+                        passed: args.len(),
+                        expected: EXPECTED,
+                    }.into())
+                }
+            }
+        }
+    };
+}
+
+// Define up to the length supported by standard library.
+varargs_into_tuple!();
+varargs_into_tuple!(A);
+varargs_into_tuple!(A, B);
+varargs_into_tuple!(A, B, C);
+varargs_into_tuple!(A, B, C, D);
+varargs_into_tuple!(A, B, C, D, E);
+varargs_into_tuple!(A, B, C, D, E, F);
+varargs_into_tuple!(A, B, C, D, E, F, G);
+varargs_into_tuple!(A, B, C, D, E, F, G, H);
+varargs_into_tuple!(A, B, C, D, E, F, G, H, I);
+varargs_into_tuple!(A, B, C, D, E, F, G, H, I, J);
+varargs_into_tuple!(A, B, C, D, E, F, G, H, I, J, K);
+varargs_into_tuple!(A, B, C, D, E, F, G, H, I, J, K, L);
+
+/// All possible error types for convert from Varargs.
+#[derive(Debug)]
+pub enum VarargsError {
+    ArgumentLengthError(ArgumentLengthError),
+    FromVariantError(FromVariantError),
+}
```

Comment:
Do we really need nested types here?

We have [struct-like enumerators for `FromVariantError`](https://github.com/godot-rust/godot-rust/blob/c69c5a0241a2206ec45e2a8028922ca10648bcfa/gdnative-core/src/core_types/variant.rs#L837) and it seems to work well.

We also don't really need all the `From<T>` implementations -- as I see it, they provide _minor_ convenience for this one file, at the cost of much more code.

Reply:
That's what I missed! :rofl:

However, `FromVariantError` does not appear to be specialized to arguments error messages. In Rust, Change of type enum is breaking change, so it cannot be changed immediately either. So it looks like we will create an `ArgumentTypeError` that is specialized for argument errors and `VarargsError` will remain nested.

Reply:
It could also be done like this, no?
```rs
#[derive(Debug)]
pub enum VarargsError {
    ArgumentLengthError  {
        passed: usize,
        expected: usize,
    },
    FromVariantError(FromVariantError),
}
```

Or if you find `ArgumentTypeError` more meaningful than `FromVariantError`, you could also do a struct-like enumerator for that.

Unless of course we find a good reason to reuse those types, but they seem used isolated.

Reply:
I did not do that, because if we write the structure directly to enum, we would not be able to write methods for each errors.